Create SafeDNS provider

rjhenry opened the pull request:
<!--
    Please read https://github.com/kubernetes-sigs/external-dns#contributing before submitting
    your pull request. Please fill in each section below to help us better prioritize your pull request. Thanks!
-->

**Description**
Create a provider for UKFast's SafeDNS.

**Checklist**

- [x] Unit tests updated
- [x] End user documentation updated

rjhenry commented:
/check-cla

rjhenry commented:
/assign raffo

rjhenry commented:
Apologies for the mess of comments on the PR - getting used to the format! Should be in the correct state for Phase 1 - human review.

rjhenry commented:
@Raffo @seanmalloy @njuettner Is there any idea on when this PR may be able to be looked at? I'm keen to get this in place so I can switch back to the upstream images for ExternalDNS.

Raffo commented:
@rjhenry we're incredibly resource constrained this month, so that might take a bit unfortunately.

rjhenry commented:
@Raffo @seanmalloy @njuettner Now that another month has passed, is there any more likelyhood of being able to review this PR, or are you still incredibly resource-constrained?

rjhenry commented:
(Note: Rebased from v0.10.0 to v0.10.1, and improved lint quality)

Raffo commented:
@rjhenry yes, we’ll take a look in a bit.

Raffo commented on docs/tutorials/UKFast_SafeDNS.md:
```diff
@@ -0,0 +1,208 @@
+# Setting up ExternalDNS for Services on UKFast's SafeDNS
+
+This tutorial describes how to setup ExternalDNS for usage within a Kubernetes cluster using SafeDNS
```
Nit: 
```suggestion
This tutorial describes how to setup ExternalDNS for usage within a Kubernetes cluster using SafeDNS.
```

Raffo commented on docs/tutorials/UKFast_SafeDNS.md:
```diff
@@ -0,0 +1,208 @@
+If you want to learn about how to use the SafeDNS service read the following tutorials:
+To learn more about the use of SafeDNS in general, see the following page:
+
+[UKFast's SafeDNS documentation](https://docs.ukfast.co.uk/domains/safedns/index.html)
```
Nit:

```suggestion
[UKFast's SafeDNS documentation](https://docs.ukfast.co.uk/domains/safedns/index.html).
```

Raffo commented on docs/tutorials/UKFast_SafeDNS.md:
```diff
@@ -0,0 +1,208 @@
+    spec:
+      containers:
+      - name: external-dns
+        image: k8s.gcr.io/external-dns/external-dns:FIXME
```
This needs edit too. Please add a comment that says that folks need to find the last version by looking at the GitHub releases as we currently still don’t have a process to replace all occurrences of the version in manifests in the doc.

Raffo commented on provider/safedns/safedns.go:
```diff
@@ -0,0 +1,239 @@
+	"sigs.k8s.io/external-dns/provider"
+)
+
+// The SafeDNS interface is a subset of the SafeDNS service API that are actually used.
```
It’s a convention to have the first word of the comment be the name of the type you are commenting. For example:

```suggestion
// SafeDNS is an interface that is a subset of the SafeDNS service API that are actually used.
```


Can you please make it consistent and make sure all exported types have comments?

Raffo commented on provider/safedns/safedns.go:
```diff
@@ -0,0 +1,239 @@
+	"os"
+
+	log "github.com/sirupsen/logrus"
+	ukf_client "github.com/ukfast/sdk-go/pkg/client"
```
I think it would be better to stick to camel case names for imports. Can you change this everywhere?

Raffo commented on provider/safedns/safedns.go:
```diff
@@ -0,0 +1,239 @@
+		}
+	}
+	for _, endpoint := range changes.UpdateNew {
+		// TODO: Find a more effient way of doing this.
+		// Currently iterates over each zoneRecord in ZoneRecords for each Endpoint in UpdateNew; the same will go for
+		// Delete. As it's double-iteration, that's O(n^2), which isn't great.
```
Should we be concerned about this? Is it something we are okay to merge like it is and then reconsider in the future?

rjhenry replied:
Personally, I haven't noticed any issues running this code in production for an account with many hundred of zones, each with many hundreds of records. On that basis alone, I'd say that it's okay to merge now and reconsider in future.

Raffo replied:
Ok sounds good. I would prefer if we leave the TODO then if we do not need any change for now, so that we can search TODOs for real things that are left to do in the codebase.

Raffo commented on provider/safedns/safedns.go:
```diff
@@ -0,0 +1,239 @@
+		}
+	}
+	for _, endpoint := range changes.Delete {
+		// TODO: Find a more effient way of doing this.
```
Same as above.

rjhenry replied:
Replied above.

Raffo reviewed (commented):
@rjhenry I added a few comments, changes look good overall. I’d be happy to approve this PR once the comments are addressed. Can you please also make sure to add yourself as a maintainer of this code in the README section? We ask for people contributing new provider to provide as well some level of support for the code they will commit.

rjhenry commented:
> Can you please also make sure to add yourself as a maintainer of this code in the README section?

I have added my organisation as a maintainer, and not myself as an individual. The organisation has a use for this, but myself as an individual does not, so it makes more sense - in my mind - for the organisation to be deemed the maintainer of this plugin.
If you'd prefer an individual, I can change this for now, though note that I'd likely avoid being a maintainer of code I don't use if I were to leave @assureddt.

Raffo reviewed (commented):
@rjhenry if you drop the TODOs, I’m ok to approve this PR.

rjhenry commented:
I've dropped the TODOs, but kept the comment about the complexity; it's likely to be a good place to look if any performance issues are reported. I've also rebased to be based on v0.10.2 rather than v0.10.1.

Many thanks for working with me on this @Raffo!

Raffo commented on docs/tutorials/UKFast_SafeDNS.md:
```diff
@@ -0,0 +1,208 @@
+
+This tutorial describes how to setup ExternalDNS for usage within a Kubernetes cluster using SafeDNS
+
+Make sure to use **>=FIXME** version of ExternalDNS for this tutorial.
```
@rjhenry can ypu replace the fixme with `0.11.0`? It will be unreleased when we merge this, but it is the next tag for release. Once you change that we're good to merge this one.

Raffo commented:
@rjhenry I have a comment to solve in https://github.com/kubernetes-sigs/external-dns/pull/2359/files#r778642557 then we can merge this PR.

rjhenry commented:
@Raffo I've removed the FIXME (and it's corresponding one in the image name, in the documentation) and fixed the lint issue; is there anything else you need from me here?

Raffo reviewed (approved):
/lgtm
